tmol/score/atom_type_dependent_term.py

```diff
@@ -29,13 +33,57 @@ def __init__(self, param_db: ParameterDatabase, device: torch.device):
         self.atom_type_index = atom_type_resolver.index
         self.np_is_hydrogen = atom_type_resolver.params.is_hydrogen.cpu().numpy()
         self.np_is_heavyatom = numpy.logical_not(self.np_is_hydrogen)
+
+        self.atom_name_index = OrderedDict()
+        self.atom_unique_id_index = OrderedDict()
         self.device = device
 
+    def get_atom_unique_id_name(self, block_name, atom_name):
+        return "UNIQUE_ID:" + block_name + ":" + atom_name
+
+    def get_atom_wildcard_id_name(self, block_name, atom_name):
+        return "WILDCARD_ID:" + atom_name
+
+    def add_atom_index(self, block_type):
+        atom_names = [x.name for x in block_type.atoms]
+
+        unique_id_indices = []
+        name_indices = []
+        for atom_name in atom_names:
+            unique_id = self.get_atom_unique_id_name(block_type.base_name, atom_name)
```

Comment:
I've been puzzling over your "quick" question from last week surrounding the issue of name collision. Here, regardless of the order in which the block types are encountered, all the atoms that will ever be encountered will be given a unique atom id. Atoms that are part of n-term patches will get one id; atoms that are part of the c-term patches will get other ids; atoms that are part of both will get the same (individually unique) ids. The order in which atoms are processed will determine which atom IDs they are given, but the actual values of the IDs themselves make no difference (besides, perhaps some small effect on performance due to differences in hash function collisions). 

In the cart-bonded code where there are two different atoms with the same name that are created by different patches of the same base type, we would have "name collision" issues, yes. I think in that case, we should instead detect when two sets of parameters are specified for the same set of atoms and provide a warning. If (:pray: @everyday847) there are two patches that both add some "CM" atom -- one on a mono-methylated n-term nitrogen, and another one a di-methylated n-term nitrogen, and let's say that the patches both try to add a (different)"wildcard-N + wildcard+CM" set of bond-length parameters, then the cart-bonded database ought to stop execution and insist that either the atom names be changed in the patches to avoid the collision, or one of the patches ought to be converted into a new set of base types.

Therefore, I think that this strategy for lookup based on base-type name is the right way to go.